File: semi-labeling/app/electron/labeling/sampling.js
// Diverse subset selection for the review queue.
//
// Goal: instead of labeling every box, pick a small, representative sample
// (~10%) that is as DISTINCT as possible. We use the cached DINOv2 tight
// embeddings and run farthest-point sampling (FPS) per class so each class is
// covered and the picks spread across the embedding space (low redundancy).

// Decode a SQLite BLOB (Uint8Array/Buffer) into a Float32Array copy.
const decodeVec = (blob) => {
  if (!blob) return null;
  const u8 = blob instanceof Uint8Array ? blob : new Uint8Array(blob);
  // copy so we don't alias a pooled buffer
  const copy = u8.slice();
  return new Float32Array(copy.buffer, copy.byteOffset, Math.floor(copy.byteLength / 4));
};

// cosine distance for L2-normalized vectors: 1 - dot
const cosineDist = (a, b) => {
  let dot = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i += 1) dot += a[i] * b[i];
  return 1 - dot;
};

// Farthest-point sampling within one class group.
// items: [{ resultId, vec: Float32Array, reliability: number }]
// returns: array of selected resultIds
const fpsSelect = (items, k) => {
  const total = items.length;
  if (k >= total) return items.map((it) => it.resultId);
  if (k <= 0) return [];

  // seed with the most-uncertain item (lowest reliability), ties by result_id
  let seedIdx = 0;
  for (let i = 1; i < total; i += 1) {
    const a = items[i];
    const b = items[seedIdx];
    if (a.reliability < b.reliability || (a.reliability === b.reliability && a.resultId < b.resultId)) {
      seedIdx = i;
    }
  }

  const selected = [seedIdx];
  const minDist = new Float64Array(total);
  for (let i = 0; i < total; i += 1) {
    minDist[i] = items[i].vec ? cosineDist(items[i].vec, items[seedIdx].vec) : Infinity;
  }
  minDist[seedIdx] = -1; // mark selected

  while (selected.length < k) {
    // pick the point with the largest distance to the selected set
    let best = -1;
    let bestDist = -Infinity;
    for (let i = 0; i < total; i += 1) {
      if (minDist[i] > bestDist) {
        bestDist = minDist[i];
        best = i;
      }
    }
    if (best < 0) break;
    selected.push(best);
    const bv = items[best].vec;
    minDist[best] = -1;
    if (bv) {
      for (let i = 0; i < total; i += 1) {
        if (minDist[i] < 0) continue;
        const d = items[i].vec ? cosineDist(items[i].vec, bv) : Infinity;
        if (d < minDist[i]) minDist[i] = d;
      }
    }
  }

  return selected.map((i) => items[i].resultId);
};

// Stratified diverse sample across class groups.
// rows: [{ resultId, label, reliability, vec }]
// ratio: 0..1 fraction to keep (per class, rounded up, min 1 per non-empty class)
// returns: Set of selected resultIds
export const selectDiverseSample = (rows, ratio) => {
  const r = Number(ratio);
  if (!Number.isFinite(r) || r <= 0 || r >= 1) {
    return new Set(rows.map((row) => row.resultId));
  }
  const byLabel = new Map();
  for (const row of rows) {
    const key = row.label || 'unknown';
    if (!byLabel.has(key)) byLabel.set(key, []);
    byLabel.get(key).push(row);
  }
  const picked = new Set();
  for (const [, items] of byLabel) {
    const k = Math.max(1, Math.round(items.length * r));
    for (const id of fpsSelect(items, k)) picked.add(id);
  }
  return picked;
};

export { decodeVec };

File: semi-labeling/semilabel_app/services/sampling.py
from __future__ import annotations

from collections import defaultdict
from math import isfinite
from array import array
from typing import Sequence

try:
    import numpy as np
except Exception:  # pragma: no cover - fallback for DB-only tests without numpy
    np = None  # type: ignore[assignment]


Vec = Sequence[float]


def decode_vec(blob: bytes | memoryview | None) -> Vec | None:
    if not blob:
        return None
    raw = bytes(blob)
    if np is not None:
        return np.frombuffer(raw, dtype=np.float32).copy()
    values = array("f")
    values.frombytes(raw)
    return values


def _cosine_dist(a: Vec, b: Vec) -> float:
    n = min(len(a), len(b))
    if n <= 0:
        return 1.0
    if np is not None:
        return float(1.0 - np.dot(a[:n], b[:n]))
    dot = 0.0
    for idx in range(n):
        dot += float(a[idx]) * float(b[idx])
    return 1.0 - dot


def _fps_select(items: list[dict], k: int) -> list[int]:
    total = len(items)
    if k >= total:
        return [int(item["result_id"]) for item in items]
    if k <= 0:
        return []

    seed_idx = 0
    for idx in range(1, total):
        current = items[idx]
        seed = items[seed_idx]
        if (
            float(current.get("reliability") or 0) < float(seed.get("reliability") or 0)
            or (
                float(current.get("reliability") or 0) == float(seed.get("reliability") or 0)
                and int(current["result_id"]) < int(seed["result_id"])
            )
        ):
            seed_idx = idx

    selected = [seed_idx]
    if np is not None:
        min_dist = np.full(total, np.inf, dtype=np.float64)
    else:
        min_dist = [float("inf")] * total
    seed_vec = items[seed_idx].get("vec")
    if seed_vec is not None:
        for idx, item in enumerate(items):
            vec = item.get("vec")
            min_dist[idx] = _cosine_dist(vec, seed_vec) if vec is not None else np.inf
    min_dist[seed_idx] = -1.0

    while len(selected) < k:
        best = int(np.argmax(min_dist)) if np is not None else max(range(total), key=lambda idx: min_dist[idx])
        if best < 0 or min_dist[best] < 0:
            break
        selected.append(best)
        best_vec = items[best].get("vec")
        min_dist[best] = -1.0
        if best_vec is None:
            continue
        for idx, item in enumerate(items):
            if min_dist[idx] < 0:
                continue
            vec = item.get("vec")
            dist = _cosine_dist(vec, best_vec) if vec is not None else np.inf
            if dist < min_dist[idx]:
                min_dist[idx] = dist

    return [int(items[idx]["result_id"]) for idx in selected]


def select_diverse_sample(rows: list[dict], ratio: float) -> set[int]:
    r = float(ratio) if isfinite(float(ratio or 0)) else 0.0
    if r <= 0 or r >= 1:
        return {int(row["result_id"]) for row in rows}

    by_label: dict[str, list[dict]] = defaultdict(list)
    for row in rows:
        by_label[str(row.get("label") or "unknown")].append(row)

    picked: set[int] = set()
    for items in by_label.values():
        k = max(1, round(len(items) * r))
        picked.update(_fps_select(items, k))
    return picked
